### PR TITLE
chore: add test for fastify-plugin usage

### PR DESCRIPTION
**Pull request by SimenB**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

I tried using `FastifyPluginAsyncTypebox` and ran into issues with `fastify-plugin`. This PR is (for now at least) just a failing test.

/cc @sinclairzx81

### CODE REVIEW

**Comment by sinclairzx81** [comment]
@SimenB Hi! Just had a quick look at this. I think the issue likely resides inside `fastify-plugin` package where the `fp` function signature doesn't seem to support varying generic plugin arguments. The current definitions for the `fp` library are linked below.

https://github.com/fastify/fastify-plugin/blob/master/plugin.d.ts#L3-L17

I'm able to resolve the issue by making the following edits to the above code. (note: these edits where made directly to the `plugin.d.ts` definition file located inside `node_modules` on this branch)

```typescript
import {
  FastifyPluginAsync,
  FastifyPluginCallback,
  FastifyPluginOptions,
  RawServerDefault,
  FastifyTypeProviderDefault
} from 'fastify'

// ---------------------------------------------------------------------------------
// Utility Types
// ---------------------------------------------------------------------------------

export type ExtractFastifyPluginAsyncParams<T> = T extends FastifyPluginAsync<
  infer Options,
  infer Server,
  infer TypeProvider
> ? {
  Options: Options,
  Server: Server,
  TypeProvider: TypeProvider
} : {
  Options: FastifyPluginOptions,
  Server: RawServerDefault,
  TypeProvider: FastifyTypeProviderDefault,
}

export type ExtractFastifyPluginCallbackParams<T> = T extends FastifyPluginCallback<
  infer Options,
  infer Server,
  infer TypeProvider
> ? {
  Options: Options,
  Server: Server,
  TypeProvider: TypeProvider
} : {
  Options: FastifyPluginOptions,
  Server: RawServerDefault,
  TypeProvider: FastifyTypeProviderDefault,
}

/**
 * This function does three things for you:
 *   1. Add the `skip-override` hidden property
 *   2. Check bare-minimum version of Fastify
 *   3. Pass some custom metadata of the plugin to Fastify
 * @param fn Fastify plugin function
 * @param options Optional plugin options
 */
export default function fp<Plugin, Params extends ExtractFastifyPluginAsyncParams<Plugin>>(fn: Plugin, options?: PluginMetadata): FastifyPluginAsync<Params['Options'], Params['Server'], Params['TypeProvider']>;
export default function fp<Plugin, Params extends ExtractFastifyPluginAsyncParams<Plugin>>(fn: Plugin, options?: string): FastifyPluginAsync<Params['Options'], Params['Server'], Params['TypeProvider']>;
export default function fp<Plugin, Params extends ExtractFastifyPluginCallbackParams<Plugin>>(fn: Plugin, options?: PluginMetadata):  FastifyPluginCallback<Params['Options'], Params['Server'], Params['TypeProvider']>;
export default function fp<Plugin, Params extends ExtractFastifyPluginCallbackParams<Plugin>>(fn: Plugin, options?: string): FastifyPluginCallback<Params['Options'], Params['Server'], Params['TypeProvider']>;

```
So, the above is a quick draft that should in theory work, but needs review from the Fastify team for overall correctness. Note, I'm not very too with the `fastify-plugin` package, but maybe the above can serve as a starting point to help get things resolved.

@RafaelGSS Any thoughts on updating the `fastify-plugin` package to support varying generics on these plugins? Maybe there's a better way to express these definitions?

Hope this helps!
S

**Comment by sinclairzx81** [comment]
https://github.com/fastify/fastify-plugin/issues/191

**Comment by RafaelGSS** [comment]
@sinclairzx81 could you confirm if the patch https://github.com/fastify/fastify-plugin/issues/191 fixes this issue?

**Comment by sinclairzx81** [comment]
@RafaelGSS Heya. Can confirm the updates on https://github.com/fastify/fastify-plugin/pull/192 do enable the TS tests on this branch to complete successfully.

File: package.json [hunk 1]
```diff
@@ -40,6 +40,7 @@
     "eslint-plugin-n": "^15.2.4",
     "eslint-plugin-promise": "^6.0.0",
     "fastify": "^4.2.0",
+    "fastify-plugin": "^4.1.0",
```
Comment (Eomm) [inline package.json:43]:
```suggestion
    "fastify-plugin": "^4.2.0",
```

**Comment by SimenB** [comment]
Sweet!

**Review by Uzlopak — APPROVED** [review]

**Pull request merged**